### PR TITLE
[GEOS-11274] Fix JSON Legend with external reference

### PR DESCRIPTION
**Pull request by iaunzu**
This a a PR for [GEOS-11274](https://osgeo-org.atlassian.net/browse/GEOS-11274)

It prevents NullPointerException providing links in the form of URI Data for unresolvable urls and excluding ``url`` and ``external-graphic-url`` when not possible.
  
# Checklist

- [X] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [X] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [X] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [X] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [X] New unit tests have been added covering the changes.
- [X] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [X] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [X] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [X] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [X] Bug fixes and small new features are presented as a single commit.
- [X] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/wms/src/main/java/org/geoserver/wms/icons/IconProperties.java [hunk 1]
```diff
@@ -166,8 +173,15 @@ public String href(String baseURL, String workspace, String styleName) {
...
-                                file = null;
+                                // other dirs are not published, serialize uri-data
+                                Path path = file.toPath();
+                                String contentType = Files.probeContentType(path);
+                                byte[] bytes = Files.readAllBytes(path);
+                                return new StringBuilder("data:")
+                                        .append(contentType)
+                                        .append(";base64,")
+                                        .append(Base64.getEncoder().encodeToString(bytes))
+                                        .toString();
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/icons/IconProperties.java:184]: Relative references should work too though, the contents of the styles directory is available to the users.
So for example, the "grass_poly.sld" [file in the release data directory](https://github.com/geoserver/geoserver/blob/main/data/release/styles/grass_poly.sld) contains this relative reference:

                    <OnlineResource xlink:type="simple" xlink:href="grass_fill.png" />

which is accessible using a link to the origin GeoServer, e.g.:

https://gs-main.geosolutionsgroup.com/geoserver/styles/grass_fill.png

Long story short, it seems to me the "file.isAbsolute" test is too strict

**Review by aaime — CHANGES_REQUESTED** [review]

**Review by aaime — CHANGES_REQUESTED** [review]

**Comment by iaunzu** [comment]
Very good point. This is indeed a serious security breach. At the very least, I should have verified that the resource is in the styles directory, just as it is checked for the main styles directory. This condition should be verified as a security check if the resource is serialized, or as a simple precaution if an URL is going to be generated.

On the other hand, I just realized that the contents of the styles directory for specific workspaces ARE published. However, the URL is not what I expected. For example, _geoserver/styles/sf/ExternalGraphicIcon.png_ does work.

I am going to push some changes with this in mind.

**Comment by aaime** [comment]
I did not notice your last change. Could you update the PR and align it with the main branch? (there are a couple of conflicts)

**Comment by iaunzu** [comment]
Done. Also fixed the test.

File: src/wms/src/test/resources/org/geoserver/wms/legendgraphic/ExternalReferenceGraphic.sld [hunk 2]
```diff
@@ -0,0 +1,28 @@
+<StyledLayerDescriptor xmlns="http://www.opengis.net/sld" xmlns:ogc="http://www.opengis.net/ogc" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (aaime) [inline src/wms/src/test/resources/org/geoserver/wms/legendgraphic/ExternalReferenceGraphic.sld:1]: While toppSample below is used, I don't see this new file being used anywhere
